src/bsoncxx/include/bsoncxx/v1/document/view.hpp

```diff
@@ -20,16 +20,407 @@
+///
+BSONCXX_ABI_EXPORT_CDECL(std::error_category const&) view();
+
+} // namespace document
+} // namespace category
+} // namespace error
+} // namespace v1
+} // namespace bsoncxx
+
+namespace bsoncxx {
+namespace v1 {
+namespace error {
+
+///
+/// Declares error codes returned by @ref bsoncxx::v1::document interfaces.
+///
+namespace document {
+
+///
+/// Errors codes which may be returned by @ref bsoncxx::v1::document::view.
+///
+/// @attention This feature is experimental! It is not ready for use!
+///
+enum class view {
+    zero,           ///< Zero.
+    invalid_length, ///< Length is invalid.
+    invalid_data,   ///< Data is invalid.
+};
+
+///
+/// Support implicit conversion to `std::error_code`.
+///
+/// @attention This feature is experimental! It is not ready for use!
+///
+inline std::error_code make_error_code(view v) {
+    return {static_cast<int>(v), v1::error::category::document::view()};
+}
+
+} // namespace document
+} // namespace error
+} // namespace v1
+} // namespace bsoncxx
+
+namespace std {
+
+template <>
+struct is_error_code_enum<bsoncxx::v1::error::document::view> : true_type {};
+
+} // namespace std
+
 namespace bsoncxx {
 namespace v1 {
 namespace document {
 
 ///
 /// A non-owning, read-only BSON document.
 ///
+/// An "invalid" view, as indicated by @ref operator bool() const, does not satisfy the minimum requirements of a valid
+/// BSON document, which are that:
+///
+/// - @ref data() is not null, and
+/// - @ref size() is not less than `5` (the minimum size of a BSON document).
+///
+/// The BSON bytes being represented is only validated as minimally required to satisfy a requested operation. When an
+/// operation is not satisfiable due to invalid data, the operation will throw an @ref bsoncxx::v1::exception with @ref
+/// bsoncxx::v1::error::document::view::invalid_data.
+///
 /// @attention This feature is experimental! It is not ready for use!
 ///
-class view {};
+class view {
+   public:
+    class const_iterator;
+
+    ///
+    /// Equivalent to @ref const_iterator.
+    ///
+    using iterator = const_iterator;
+
+   private:
+    enum : std::size_t { _empty_length = 5u };
+
+    std::uint8_t const* _data;
+
+    template <detail::endian e = detail::endian::native>
+    std::int32_t raw_size() const;
+
+   public:
+    ///
+    /// Initialize as an empty view.
+    ///
+    /// @par Postconditions:
+    /// - `this->data() != nullptr`
+    /// - `this->size() == 5`
+    /// - `this->empty() == true`
+    ///
+    BSONCXX_ABI_EXPORT_CDECL() view();
+
+    ///
+    /// Initialize with the given BSON bytes.
+    ///
+    /// @par Preconditions:
+    /// - If `data` is not null, the size of the storage region pointed to by `data` must be greater than or equal to 5.
+    /// - The embedded length must be less than or equal to the size of the storage region pointed to by `data`.
+    ///
+    explicit view(std::uint8_t const* data) : _data{data} {}
```

Comment:
Consider: Use a named-constructor idiom to call out the unsafe-ness of this function:

```c++
auto v = bsoncxx::view::static_function_with_conspicuous_name_that_creates_the_object_without_checking(data_ptr);
```